applications/ray/kuberay-tpu-webhook/main.go

```diff
@@ -107,57 +119,89 @@ func extractRayCluster(admissionReview *admissionv1.AdmissionReview) (*ray.RayCl
-	containers := workerGroupSpec.Template.Spec.Containers
-	if containers == nil {
-		klog.Fatalf("Container path not specified")
-	}
-	// inject subdomain and TPU_WORKER_HOSTNAMES into pods for TPU multi-host initialization
-	for j := 0; j < len(containers); j++ {
-		container := containers[j]
-		if containerRequestingTPUs(container) {
-			subdomainPatch, hostNamesPatch := patch{"op": "add"}, patch{"op": "add"}
-			subdomainPath := fmt.Sprintf("/spec/workerGroupSpecs/%d/template/spec/subdomain", workerGroupIndex)
-			envPath := fmt.Sprintf("/spec/workerGroupSpecs/%d/template/spec/containers/%d/env", workerGroupIndex, j)
-			tpuWorkerHostNames := corev1.EnvVar{
-				Name:  "TPU_WORKER_HOSTNAMES",
-				Value: hostNames,
-			}
-			subdomainPatch["path"] = subdomainPath
-			subdomainPatch["value"] = headlessServiceName
-			// create new EnvVar array if container.Env is empty, and append hostnames if not
-			if len(container.Env) == 0 {
-				hostNamesPatch["path"] = envPath
-				hostNamesPatch["value"] = []corev1.EnvVar{tpuWorkerHostNames}
-			} else {
-				hostNamesPatch["path"] = fmt.Sprintf("%s/-", envPath)
-				hostNamesPatch["value"] = tpuWorkerHostNames
-			}
-			*patches = append(*patches, subdomainPatch, hostNamesPatch)
-		}
-	}
+// inject subdomain and TPU_WORKER_HOSTNAMES into pods for TPU multi-host initialization
+func injectHostnames(hostNames string, envPath string, container corev1.Container, patches *[]patch) {
+	subdomainPatch, hostNamesPatch := patch{"op": "add"}, patch{"op": "add"}
+	subdomainPath := "/spec/subdomain"
+	tpuWorkerHostNames := corev1.EnvVar{
+		Name:  "TPU_WORKER_HOSTNAMES",
+		Value: hostNames,
+	}
+	subdomainPatch["path"] = subdomainPath
+	subdomainPatch["value"] = headlessServiceName
+	// create new EnvVar array if container.Env is empty, and append hostnames if not
+	if len(container.Env) == 0 {
+		hostNamesPatch["path"] = envPath
+		hostNamesPatch["value"] = []corev1.EnvVar{tpuWorkerHostNames}
+	} else {
+		hostNamesPatch["path"] = fmt.Sprintf("%s/-", envPath)
+		hostNamesPatch["value"] = tpuWorkerHostNames
+	}
+	*patches = append(*patches, subdomainPatch, hostNamesPatch)
+}
+
+func injectMultiHostReplicaLabel(replicaIndex int, workerGroupName string, patches *[]patch) {
+	labelPatch := patch{"op": "replace"}
+	labelPath := "/metadata/labels/multiHostReplica"
+	multiHostReplicaValue := workerGroupName + "-" + strconv.Itoa(replicaIndex)
+
+	labelPatch["path"] = labelPath
+	labelPatch["value"] = multiHostReplicaValue
+
+	*patches = append(*patches, labelPatch)
+}
+
+// inject pod affinity and anti-affinity scheduling constraints using multiHostReplica label
+func injectPodAffinity(replicaIndex int, workerGroupName string, patches *[]patch) {
+	podAffinityPatch := patch{"op": "add"}
+	podAffinityPath := "/spec/affinity"
+	podAntiAffinityPatch := patch{"op": "add"}
+	podAntiAffinityPath := "/spec/affinity"
```

Comment:
/spec/antiaffinity?

Reply:
The path is `/spec/affinity/podAffinity` and `/spec/affinity/podAntiAffinity` for podAffinity and podAntiAffinity, I'll remove the duplicate var though and use `podAffinityPath` for both patches.